lib/garbage-deleter.js

```diff
@@ -0,0 +1,893 @@
+                objectId: fields[2]
+            },
+            'Deleted one object.'
+        );
+
+        self.addCounter('deleteCountTotal', 1);
+        self.addCounter('deleteTimeSeconds', elapsed);
+        curval = self.getGauge('deleteTimeMaxSeconds');
+        if (curval === undefined || elapsed > curval) {
+            self.setGauge('deleteTimeMaxSeconds', elapsed);
+        }
+        curval = self.getGauge('deleteTimeMinSeconds');
+        if (curval === undefined || elapsed < curval) {
+            self.setGauge('deleteTimeMinSeconds', elapsed);
+        }
+
+        if (unlinkErr) {
+            if (unlinkErr.code === 'ENOENT') {
+                self.log.debug(
+                    {
+                        filename: deleteFile
+                    },
+                    'File did not exist. Skipping.'
+                );
+                self.addCounter('deleteCountMissing', 1);
+            } else {
+                callback(
+                    new VError(
+                        {
+                            cause: unlinkErr,
+                            info: {
+                                filename: deleteFile
+                            },
+                            name: 'UnlinkFileError'
+                        },
+                        'Failed to delete file.'
+                    )
+                );
+                self.addCounter('deleteErrorCount', 1);
+                return;
+            }
+        } else {
+            self.addCounter('instructionFilesDeleted', 1);
+        }
+
+        callback();
+    });
+};
+
+GarbageDeleter.prototype.processFile = function processFile(
+    instrFile,
+    callback
+) {
+    var self = this;
+
+    var badFilename = path.join(self.badInstructionDir, instrFile);
+    var beginning = process.hrtime();
+    var filename = path.join(self.instructionDir, instrFile);
+    var lineCount = 0;
+
+    // We use this function so that we can add a hook for tests to be able to
+    // know the result from each file that was processed.
+    function _doneProcessing(err) {
+        self.addCounter('instructionFilesProcessed', 1);
+
+        // On any error, we move the instruction file to the bad_instructions
+        // directory.
+        if (err) {
+            self.addCounter('instructionFilesBad', 1);
+
+            self.log.warn(
+                {
+                    errInfo: VError.info(err),
+                    filename: filename
+                },
+                'Failed to process file, moving to bad_instructions.'
+            );
+
+            self.fsRename(filename, badFilename, function _onRename(renameErr) {
+                if (renameErr) {
+                    self.log.error(
+                        {
+                            err: renameErr,
+                            srcFilename: filename,
+                            targFilename: badFilename
+                        },
+                        'Failed to rename file to bad_instructions.'
```

Comment:
Not sure anything should change here, but IIUC this'll result in a `self.maxRunWait` infinite looping over that instruction file if for some reason it can't be removed. By default that's re-processing that file every 10s. In the limit that could be 1000 attempted no-op fs.unlink's.  Which is better: doing that, or crashing the garbage-deleter?

Reply:
I think it's better to keep looping since we're still going to be processing all the other entries. If we crash, we'll crash again as soon as we start and quickly go into maintenance and not be processing anything. In this scenario the `gc_storage_instruction_files_bad_count_total` metric will also be incremented for each loop so hopefully those in charge of monitoring will notice and be able to look into the problem.

Reply:
Cool. Sounds good.